draft-ietf-taps-interface.md

```diff
@@ -1780,37 +1772,179 @@ that the other endpoint is done sending on a connection.
+to the connection that it is ready to handle sending Messages by calling `ready` on
+the FramerInstance. This call can be delayed in case the protocol needs to do some
+simple handshake exchange before handling application Messages.
+
+~~~
+// instance is a FramerInstance reference
+instance.ready()
+~~~
+
+At any time if the Message Framer encounters a fatal error, it can also cause the Connection
+to fail and provide an error.
+
+~~~
+instance.failed(Error)
+~~~
+
+Before a Message Framer marks itself as ready, it can also dynamically add a protocol or framer
+above it in the stack. This allows protocols like STARTTLS, that need to add TLS conditionally,
+to modify the Protocol Stack based on a handshake result.
+
+~~~
+otherFramer := NewMessageFramer()
+instance.prependProtocol(otherFramer)
+~~~
+
+## Sender-side Message Framing {#send-framing}
+
+Message Framers define an event that gets called on an instance of a
+Framer to handle a new Message being sent by the application.
+
+~~~
+FramerInstance -> NewSentMessage<messageData, messageContext, endOfMessage>
+~~~
+
+Upon receiving this event, a Message Framer implementation is responsible for
+performing any necessary transformations and sending the resulting data to the next
+protocol. Implementations SHOULD ensure that there is a way to pass the original data
+through without copying to improve performance.
+
+~~~
+instance.send(Data)
+~~~
+
+To provide an example, a simple protocol that adds a length as a header would receive
+the `NewOutputMessage` event, create a data representation of the length of the Message
+data, and then send a block of data that is the concatenation of the length header and the original
+Message data.
+
+## Receiver-side Message Framing {#receive-framing}
+
+In order to parse an received flow of data into Messages, the Message Framer
+is first notified whenever new data is available to parse.
+
+~~~
+FramerInstance -> HandleReceivedData<>
+~~~
+
+Upon receiving this event, a Message Framer can inspect the inbound data. The
+data is parsed from a particular cursor representing the unprocessed data. The
+Framer requests a specific amount of data it needs to have available in order to parse.
+If the data is not available, the parse fails.
+
+~~~
+instance.parse(minIncompleteLength, maxLength) -> (Data?, endOfMessage?)
+~~~
+
+The Message Framer can directly advance the receive cursor once it has
+parsed data to effectively discard data (for example, discard a header
+once the content has been parsed).
+
+To deliver a Message to the application, the Framer can either directly
+send data that it has created, or deliver a range of data directly from the underlying
+transport and simulatenously advance the receive cursor.
+
+~~~
+instance.advanceReceiveCursor(length)
+instance.deliverAndAdvanceReceiveCursor(messageContext, length, endOfMessage)
+instance.deliver(messageContext, data, endOfMessage)
+~~~
```

Comment:
This is another point where I am unsure on how the framer actually interfaces with the connection. Who calls instance.advanceReceiveCursor?

Reply:
The application's implementation of the framer.